include/glow/Support/Error.h

```diff
@@ -16,40 +16,219 @@
+  } while (0)
 
-/// Is true_type only if applied to llvm::Expected.
-template <typename> struct IsLLVMExpected : public std::false_type {};
+/// Takes an Error and returns it if it's not success.
+// TODO: extend this to work with Expected as well.
+#define RETURN_IF_ERR(err)                                                     \
+  do {                                                                         \
+    if (auto errV = std::forward<Error>(err)) {                                \
+      static_assert(glow::detail::IsError<decltype(errV)>::value,              \
+                    "Expected value to be a Error");                           \
+      return std::forward<Error>(errV);                                        \
+    }                                                                          \
+  } while (0)
+
+/// Takes an Error and if it contains an ErrorValue then calls FAIL().
+#define FAIL_TEST_IF_ERR(err)                                                  \
+  do {                                                                         \
+    if (auto errV = std::forward<Error>(err)) {                                \
+      static_assert(glow::detail::IsError<decltype(errV)>::value,              \
+                    "Expected value to be a Error");                           \
+      FAIL() << errorToString(std::move(errV));                                \
+    }                                                                          \
+  } while (0)
+
+/// Takes a predicate \p and if it is false then creates a new Error
+/// and returns it.
+#define RETURN_ERR_IF_NOT(p, ...)                                              \
+  do {                                                                         \
+    if (!(p)) {                                                                \
+      RETURN_ERR(__VA_ARGS__);                                                 \
+    }                                                                          \
+  } while (0)
+
+/// Forward declarations.
+namespace detail {
+class GlowError;
+class GlowErrorSuccess;
+class GlowErrorEmpty;
+class GlowErrorValue;
+template <typename T> class GlowExpected;
+} // namespace detail
+
+/// Type aliases to decouple Error and Expected from underlying implementation.
+using Error = detail::GlowError;
+using ErrorSuccess = detail::GlowErrorSuccess;
+using ErrorEmpty = detail::GlowErrorEmpty;
+using ErrorValue = detail::GlowErrorValue;
+template <typename T> using Expected = detail::GlowExpected<T>;
+
+/// NOTE: detail namespace contains code that should not be used outside of
+/// Error.h and Error.cpp. Please instead use types and macros defined above.
+namespace detail {
+/// enableCheckingErrors is used to enable assertions that every Error and
+/// Expected has its status checked before it is destroyed. This should be
+/// enabled in debug builds but turned off otherwise.
```

Comment:
BTW, a quick question: What if this could be controlled by a command-line flag instead? It would be disabled by default in release builds, but one could enable it via such a command-line option if needed? Would it bring any value?

Reply:
That would be worse because when it is disabled at build time
1. Error and Expected use less space because no need to store `CheckState:: checked_`
2. The code could be slightly faster because the code to do the checking is completely non-existent so no need even to check a flag to see if the state should be checked.
Also this is really a debugging mechanism so I think it makes sense to enable it in only debug mode. My view on it is that ideally, ensuring that Errors are always checked would be done statically but this is Rust so having some debug assertions is probably the next best thing, but we shouldn't be putting that check into Release. What do you think?

Reply:
Yeah, makes sense. Thanks for the explanation!